### ISSUE TITLE
"takes widget" does not return TriggerRegisterDeathEvent 

### ISSUE THREAD

**Issue by Luashine** [lep/jassdoc-browser#2]
https://lep.duckdns.org/app/jassbot/search?query=takes+widget does not return [TriggerRegisterDeathEvent](https://lep.duckdns.org/app/jassbot/doc/TriggerRegisterDeathEvent) although it takes widget too. Weird.

**Comment by lep** [comment]
Yes it's weird; and also no quick fix available I think. Currently it just so happens that numbers in jassbot favour a type substitution with a related type to adding another parameter. Once I come around rewriting the jassbot search I will definitively keep this example in mind.

It also depends a bit on what the user expects. Should a `takes widget` favour functions which take unit or which take handle? I assume we could model it on [co- and contra-variance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) in the future (but only favour and not exclude).

**Comment by Luashine** [comment]
I'll utter a few words you'll like (i know you will)

>  Should a takes widget favour functions which take unit or which take handle?

Following graph theory, handle → agent → widget → unit. So widget is directly connected to both agent and unit, they should take precedence over handle, because they're closest by path. Alternatively you can specify that a more specific node is to be prioritized over least specific node: unit > agent.

Based on what I would expect and the point of the search engine:

1. Show all literal `takes widget`
2. (debatable) show agent/unit
3. Handle at the very bottom

> we could model it on [co- and contra-variance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) in the future

Type system ⊂ graph theory :nerd_face: 

PS: Yeah I've described everything you said in my own words :blush:

**Comment by lep** [comment]
> Following graph theory, handle → agent → widget → unit. So widget is directly connected to both agent and unit, they should take precedence over handle, because they're closest by path. Alternatively you can specify that a more specific node is to be prioritized over least specific node: unit > agent.

I already take that into account as I compare the path to a potential common ancestor. It's just that the numbers are not favouring the `takes widget, trigger` function.

I did end up changing the search function in here [jassbot#69f868](https://github.com/lep/jassbot/commit/69f8687a15701a8d8c8b30c10c564048f31e9df0). From my quick tests the search results didn't deteriorate too much but let me know if there are any very wrong results.
